### DISCUSSION TITLE
아이템 8. finalizer와 cleaner 사용을 피하라

### DISCUSSION THREAD

**Discussion by Irisation23** [Study-2-Effective-Java/Effective-Java#22] [3. 과제] [answered]
> 🚨 들어가기에 앞서 `객체 소멸자`란 무엇인지에 대해 알아보자.

# 🤔 객체 소멸자란?
- 소멸자는 객체의 수명이 끝나면(객체의 참조가 끝나면) 컴파일러에 의해 자동으로 호출되며, 사용이 끝난 객체를 정리해 준다.
- 소멸자는 인수(인자)를 가지지 않는다.
- 소멸자는 반환값이 없지만 void형으로 선언하지 않는다.
- 객체는 여러개의 생성자를 가질 수 있지만, 소멸자는 단 하나만 가질 수 있다.

하지만 이는 자바에서 생소한 개념이다.
자바에서는 객체의 참조가 끝나면 GC(Garbage Collector)가 활동을 해 객체를 수거 해간다.
finalizer와 cleaner는 자바에서 사용되는 객체 소멸자이다.

# 🤔 finalizer 메서드란?
- 자원 누수를 방지하기 위해 JVM(Java Virtual Machine)이 실행하는 가비지 컬렉션이 수행 될 때 더 이상 사용하지 않는 자원에 대한 정리 작업을 진행하기 위해 호출되는 종료자 메서드이다.
- 이는 자바의 모든 클래스 최상위 클래스인 Object 클래스의 메서드 중 하나이다.
- JVM은 객체당 한 번만 `finalize()`메서드 호출을 허용한다.

# 🤔 cleaner 메서드란?
- Java 9에서는 finalizer가 deprecated 됐고 cleaner가 새로 생겼다.
- cleaner는 별도의 쓰레드를 사용해서 finalizer보다는 덜 위험하지만, 여전히 예측불가하고, 느리며 불필요하다.

> 🚨 GC 의 활동 범주
> 1. 익명 객체: 익명 객체는 참조 변수 없이 생성된다. 결과적으로 익명 객체를 만든 후에는 익명 객체를 엑세스 할 수없다.
> 2. Nulling 참조: 객체가 null 값을 참조하는 경우 null 값만 보유하므로 참조되지 않은 객체로 간주된다.
> ``` java
> Foo foo = new Foo();
> foo = null;
> ```
> 3. 다른 객체에 대한 참조 할당: 한 객체가 다른 객체에 할당되면 첫 번째 객체는 쓸모가 없으며 참조를 잃은 객체는 쓰레기(가비지 콜렉터의 수거 대상)이 된다.
> ```java
> Foo s1 = new Foo();
> Foo s2 = new Foo();
> s1 = s2;
> ```

가비지 컬렉터는 JVM의 일부이다. 가비지 컬렉터는 모든 객체가 JVM에 의해 저장되는 힙 메모를 확인하여 더이상 필요하지 않은 참조되지 않은 객체를 찾는다. 그리고 해당 객체를 자동으로 파괴한다. 가비지 컬렉터는 객체를 소멸시키기 전에 활동 정리를 위해 `finalize()` 메서드를 호출한다.

### 그렇다면 `정리활동`이란 무엇일까?
정리 작업은 객체가 소멸되기 전에 객체에서 사용중인 모든 리소스를 닫는 프로세스이다.
모든 객체에서 사용할 수 있는 리소스는 데이터베이스 연결, 네트워크 연결 등이 있다.
이러한 리소스를 해제하고 GC는 정리 작업을 수행한 다음 객체를 삭제한다.

### 📚 1차 정리
- 가비지 컬렉터는 항상 Java에서 `finalize()` 메서드를 호출하여 객체를 파괴하기 전에 정리 작업을 수행한다.
- `finalize()` 메서드를 사용하여 정리활동을 수행하는 이 프로세스를 종료라고 한다.

### ✨ 자바에서의 객체 수명
![image](https://user-images.githubusercontent.com/53285909/209359828-ce1c88f8-dcd8-4c9f-bec6-30f21fe14973.png)

# 🔥 문제점
### (1) finalizer와 cleaner는 제때 실행되어야 하는 작업을 보장하지 못한다. 
  - e.g) 파일 닫기에 finailzer를 사용한다 가정한다면 시스템이 동시에 열 수 있는 파일 개수가 한정되어있다.
  - 시스템이 finalizer의 실행을 게을리해서 파일을 열어 둔다면 새로운 파일을 열지 못해 프로그램이 실패 할 수있다.
### (3) 상태를 영구적으로 수정하는 작업에서는 절대 finalizer나 cleaner에 의존해서는 안된다.
  - e.g) 데이터베이스 같은 공유 자원의 영구 락(lock) 해제를 finalizer나 cleaner에 맡겨 놓으면 분산 시스템 전체가 서서히 멈출 것이다.
### (4) finalizer와 cleaner는 심각한 성능 문제도 야기한다.
 -  객체가 필요 없어진 시점 - finalize(또는 cleaner)가 실행되는 시점까지 얼마나 소요될지 예측이 어렵다.
 - GC의 효율을 떨어트리게 되므로 객체 생성, 소멸 시간이 많게는 수백배까지 차이가 날 수 있다.

### (5) finalizer를 사용한 클래스는 finalizer 공격에 노출되어 보안 문제가 야기된다.
  - 생성자나 직렬화 과정에서 예외가 발생하면, 이 생성 되다 만 객체에서 악의적인 하위 클래스의 finalizer가 수행 될 수있다. 
  - [간단한 참조 예제](https://yangbongsoo.tistory.com/8?category=919799)
### (6) ⭐️ JVM에는 가비지 컬렉터의 프로세스를 가지고 있고 GC 호출에 대해 더 잘 알고있다.

---

# ✨ 사용하는 경우

### (1) 안전망 역할
- 자원의 소유자가 `close()`를 호출하지 않는 것에 대비하는 경우
- 늦게나마 회수하는 것이 회수하지 않는 것 보다 낫기 때문

### (2) `네이티브 피어`와 연결된 객체
> 🤔 네이티브 피어란?
> [해당 링크에서 david215의 정리 참조](https://github.com/java-squid/effective-java/issues/8)
> 네이티브 피어는 일반 자바객체가 네이티브 메서드를 통해 기능을 위임한 네이티브 객체를 말한다. 자바 객체가 아니므로 GC는 네이티브 객체의 존재조차 모른다.
> ```java
> public class HelloJNI {
>
>    static {
>        System.loadLibrary("native");
>    }
>
>    public static native void helloFromC();
>
>    public static void main(String[] args) {
>        helloFromC();
>    }
> }
> ```

- C/C++ 이나 어셈블리 프로그램을 컴파일한 기계어 프로그램이자, GC가 존재를 알지 못하는 리소스
- 단, 성능 저하를 감당할 수 있고, 네이티브 피어가 심각한 자원을 가지고 있지 않는 경우에 사용.

> 📚 핵심 정리
> - cleaner(자바 8까지는 finalizer)는 안전망 역할이나 중요하지 않은 네이티브 자원 회수 용으로만 사용하자
> - 그러나 이 경우도 불확실성과 성능 저하에 주의하자.
> - [@coalong 님이 정리 해 주신 AutoCloseable](https://github.com/orgs/Study-2-Effective-Java/discussions/28)을 구현해서 사용하자.

**Comment by Irisation23** [comment, marked as answer]
@chikeem90 님이 말씀해 주신대로 해석에 오류가 있을 수 있다고 판단 되어서 작게나마 정리해 봅니다.

> - 객체가 필요 없어진 시점 - finalizer(또는 cleaner)가 실행되는 시점까지 얼마나 소요될지 예측이 어렵다.
> - GC의 효율을 떨어트리게 되므로 객체 생성, 소멸 시간이 많게는 수백배까지 차이가 날 수 있다.
  > - GC의 효율이 떨어지게 되는 이유로는 `finalize()` 를 메서드를 강제하게 된다면 GC의 객체 수거 알고리즘이 어긋날 수 있다는 생각이 듭니다.

> - (6) ⭐️ JVM에는 가비지 컬렉터의 프로세스를 가지고 있고 GC 호출에 대해 더 잘 알고있다.
> - GC 프로세스를 건드리게 되는것이 바로 `finalize()` 를 오버라이드 하게되는 경우라고 판단이 됩니다. 그리고 과연 객체 수거 시점을 보장할 수 없는 해당 메서드의 행위가 가비지 컬렉터의 프로세스가 동작하는 방식보다 빠르다는 것을 보장 할 수 없다 정도로 이해 했습니다.

**Reply by chikeem90** [reply]
@bunsung92 님
오,, 저도 좀 서칭해보고 다시 얘기해보려고,, 코멘트 지웠는데,, ㅎㅎ; 
정리 감사합니다

**Reply by chikeem90** [reply]
https://stackoverflow.com/a/2860192

stackoverflow에도 유사하게 써있네요

**Reply by Irisation23** [reply]
https://github.com/orgs/cs-collections/discussions/12

해당 부분이 GC를 조금 연구한 부분인데 혹시 GC에 대한 `작은(?) 궁금증`이 생긴다면 같이 읽어보면 도움이 되지 않을까 남겨봅니다.

**Comment by JoisFe** [comment]
참 GC만 나오면 어렵네요 ㅎㅎ
#15 `다 쓴 객체 참조를 해제하라` 라는 주제에서 다 쓴 객체를 null로 두는 이유가
GC의 활동 범주의 2.Nulling 참조: 객체가 null 값을 참조하는 경우 null 값만 보유하므로 참조되지 않은 객체로 간주된다.
때문이군요..

아이템 뒷부분을 읽으면 이 전 아이템이 더 이해가 잘 되는 부분이 있는 것 같아요.

또한 finalizer, cleaner에 대해 거의 알지 못하였는데 이번 정리 내용을 통해 알게되어 매우 유익했습니다.
감사합니다.
